Raise AttributeError in _OpsNamespace if __self__ attribute is requested

tarun292 opened the pull request:
Summary:
Trying to get the `__self__` attribute on any `_OpNamespace` object should be an invalid operation. The `__self__` attribute only exists on instance method object and not on class objects.

In [dynamo](https://github.com/pytorch/pytorch/blob/a152b3e3b8a5f543634fad07f95ef6793c78ae53/torch/_dynamo/variables/torch.py#L164) there is code that tries to access the `__self__` attribute on `TorchVariable`, this currently results in an expensive call to `torch._C._jit_get_operation` [here](https://github.com/pytorch/pytorch/blob/a152b3e3b8a5f543634fad07f95ef6793c78ae53/torch/_ops.py#L740) which ultimately fails and throws an exception. For cases where it fails the operation turns out to be quite expensive on the order of ~0.03s.

For edge use cases when exporting large models with quantized ops this exception is thrown 100's of times resulting in a lot of time wasted. By preventing the call to `torch._C._jit_get_operation` we can quickly return from this function and significantly reduce export times. On a large ASR model for example export currently takes **~405** seconds. With this change we can reduce it to **~340s**.

Overall this should also be a harmless change as no one should mostly ever try to access the `__self__` attribute on any `_OpNamespace` object.

Test Plan: Added test case.

Differential Revision: D46959879

tarun292 commented:
/easycla

larryliu0820 reviewed (approved):
I think this is fine as long as CI jobs are happy.

iseeyuan reviewed (commented):
Nice to have 1min faster export time :)

zou3519 commented on test/jit/test_custom_operators.py:
```diff
@@ -41,6 +41,16 @@ def test_dynamic_op_registry(self):
         self.assertEqual(op, op2)
 
+    def test_getting_invalid_attr(self):
+        torch.ops._test
```
Is this line necessary?

tarun292 replied:
Good point, not needed. Got rid of it.

tarun292 committed:
Raise AttributeError in _OpsNamespace if __self__ attribute is requested (#104096)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/104096

Trying to get the `__self__` attribute on any `_OpNamespace` object should be an invalid operation. The `__self__` attribute only exists on instance method object and not on class objects.

In [dynamo](https://github.com/pytorch/pytorch/blob/a152b3e3b8a5f543634fad07f95ef6793c78ae53/torch/_dynamo/variables/torch.py#L164) there is code that tries to access the `__self__` attribute on `TorchVariable`, this currently results in an expensive call to `torch._C._jit_get_operation` [here](https://github.com/pytorch/pytorch/blob/a152b3e3b8a5f543634fad07f95ef6793c78ae53/torch/_ops.py#L740) which ultimately fails and throws an exception. For cases where it fails the operation turns out to be quite expensive on the order of ~0.03s.

For edge use cases when exporting large models with quantized ops this exception is thrown 100's of times resulting in a lot of time wasted. By preventing the call to `torch._C._jit_get_operation` we can quickly return from this function and significantly reduce export times. On a large ASR model for example export currently takes **~405** seconds. With this change we can reduce it to **~340s**.

Overall this should also be a harmless change as no one should mostly ever try to access the `__self__` attribute on any `_OpNamespace` object.

Test Plan: Added test case.

Reviewed By: ezyang, zou3519

Differential Revision: D46959879

fbshipit-source-id: 6c057d01c2b745d49961136cbd24e5c72a7641fb

tarun292 commented:
@pytorchbot merge

tarun292 commented:
@pytorchbot label "topic: not user facing"

tarun292 commented:
@pytorchbot merge
